File: core/src/io/github/michelfaria/breadprototype/strategy/BlockSpawner.java
package io.github.michelfaria.breadprototype.strategy;

import com.badlogic.gdx.scenes.scene2d.Actor;
import com.badlogic.gdx.scenes.scene2d.Stage;
import io.github.michelfaria.breadprototype.actor.Block;
import io.github.michelfaria.breadprototype.actor.BlockFactory;
import io.github.michelfaria.breadprototype.actor.DirtBlock;
import io.github.michelfaria.breadprototype.actor.TntBlock;

public class BlockSpawner {

    private final Stage stage;
    private final BlockFactory<DirtBlock> dirtBlockBlockFactory;
    private final BlockFactory<TntBlock> tntBlockBlockFactory;

    //temp
    public boolean spawnDirt = false;

    public BlockSpawner(Stage stage, BlockFactory<DirtBlock> dirtBlockBlockFactory, BlockFactory<TntBlock> tntBlockBlockFactory) {
        this.stage = stage;
        this.dirtBlockBlockFactory = dirtBlockBlockFactory;
        this.tntBlockBlockFactory = tntBlockBlockFactory;
    }

    public Block spawnBlock(float x, float y) {
        Block block;
        if (spawnDirt) {
            block = dirtBlockBlockFactory.make();
        } else {
            block = tntBlockBlockFactory.make();
        }
        block.init();
        stage.addActor(block);
        block.setPosition(x, y);
        block.addNewCreationEffect();
        return block;
    }

    public void removeBlock(float x, float y) {
        final Actor hit = stage.hit(x, y, true);
        if (hit != null) {
            if (hit instanceof Block) {
                ((Block) hit).dispose();
            }
        }
    }
}

File: core/src/io/github/michelfaria/breadprototype/strategy/BlockSpawner.kt
package io.github.michelfaria.breadprototype.strategy

import com.badlogic.gdx.graphics.g2d.ParticleEffectPool
import com.badlogic.gdx.graphics.g2d.TextureAtlas
import com.badlogic.gdx.physics.box2d.World
import com.badlogic.gdx.scenes.scene2d.Stage
import io.github.michelfaria.breadprototype.actor.Block

class BlockSpawner(private val stage: Stage,
                   private val world: World,
                   private val atlas: TextureAtlas,
                   private val blockCreationEffectPool: ParticleEffectPool) {

    fun spawnBlock(x: Float, y: Float): Block {
        return Block(world, atlas, blockCreationEffectPool).also {
            stage.addActor(it)
        }.apply {
            setPosition(x, y)
            addNewCreationEffect()
        }
    }

    fun removeBlock(x: Float, y: Float) {
        stage.hit(x, y, true)?.apply {
            if (this is Block) {
                kill()
            }
        }
    }
}
